### PR TITLE
Move at::chunk into the graph fuser

### PR DESCRIPTION
**Pull request by zou3519**
... to avoid slow at::chunk (it is slow due to tensor initialization). Picking up from #10026

This is done through the following:

1) Absorb starting chunks into FusionGroup as a part of the graph fuser
pass.
2) When compiling a kernel, emit a `std::vector<ConcatDesc>` that describes if an input (of the original graph) will be chunked.
3) When launching a kernel, `use std::vector<ConcatDesc>` to chunk an
input tensor on the CPU. This chunk directly takes in an at::Tensor and creates
four TensorInfo structs in-place in the argument list, bypassing the creation of intermediate Tensors.

## Test Plan

- Expect test and correctness test to see if a single chunk is fused
  by the graph fuser
- Correctness test for a variety of chunks (dimension = beginning,
  middle, end) and tensors (contiguous, non-contiguous, edge case
  (splitSize = 1) for both CPU/CUDA
- Expect test for multiple chunks fused into the same kernel and
  correctness test.

cc @zdevito @apaszke 

## Perf
LSTM forward pass, 1 layer, 512 hidden size and input size, 100 seq length, requires_grad=False on all inputs and weights.

After changes:
```
thnn    cudnn   jit
8.8468  6.5797  9.3470
```

Before changes:
```
thnn    cudnn   jit
9.9221  6.6539  11.2550
```

### CODE REVIEW

File: torch/csrc/jit/fusion_compiler.h [hunk 1]
```diff
@@ -60,6 +60,15 @@ struct AnnotatedGraph {
   std::vector<TensorDesc> output_desc;
 };
 
+struct ChunkDesc {
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.h]: Now that I think about this more, I believe ChunkDesc and ConcatDesc are actually exactly the same thing. Each describes how to split up a bigger tensor into smaller components that become inputs into the kernel. The only difference is that for chunks it describes an input, and for cat, an output. However from the kernels perspective they are all just pointers to memory anyway. It should cleaner to actually have the code reflect this similarity, and because they are the same concept, they should also be able to share a lot of code. For instance I was suggesting adding a precomputed TensorDesc to Chunk -- well concat already has that and the handling for it.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 2]
```diff
@@ -624,27 +641,179 @@ void compressContiguous(
...
+// XXX: assumes graph inputs are 32-bit indexable
+// assumes v is an input to the graph.
+static size_t toInputIndex(Value * v, Graph & subgraph) {
+  auto start = subgraph.inputs().begin();
+  auto it = std::find(start, subgraph.inputs().end(), v);
+  JIT_ASSERT(it != subgraph.inputs().end());
+  return it - start;
+}
+
+// If there are fused Chunk nodes in the graph, returns a new Graph replacing
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.cpp]: If you look at how this is handled for concat, it takes a more lightweight approach. By leaving the Concat operators in and just skipping them in code generation we can still generate the correct kernel, and make sure the input list has the 'flattened' inputs added, without having to copy/do surgery on the fusion graph. It is possible that something similar might work here (see later comment about how chunk and concat are basically the same exact operator).

File: torch/csrc/jit/fusion_compiler.cpp [hunk 3]
```diff
@@ -728,6 +909,13 @@ struct CUDAFusionFunction : public CompiledFusionFunction {
     TORCH_CUDA_CHECK(cudaGetDeviceProperties(&prop, agraph.device));
     checkCUDAVersion(prop);
 
+    // Use a graph without the fused chunk nodes to emit a kernel.
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.cpp]: It doesn't seem right for implementation details like having fused chunks to leak into the CUDA/CPU backend code and get repeated twice. This will be hard to maintain if every new feature added to the fuser had to do something similar.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 4]
```diff
@@ -339,7 +339,11 @@ struct GraphFuser {
           // cases we inline the constants directly in the body of the fused group.
           JIT_ASSERT(input->node()->kind() == prim::Constant);
           Node * in_const = subgraph.createClone(input->node(), [](Value*) -> Value* { throw std::runtime_error("unexpected input"); });
-          subgraph.prependNode(in_const);
+          if (insert_after == nullptr) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/graph_fuser.cpp]: what caused this to need to change?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 5]
```diff
@@ -349,17 +353,24 @@ struct GraphFuser {
...
     // remapping nodes that used the input to the newly-merged node
     // n is not an input when the fusion group is empty
     auto inputs = group->inputs();
-    auto it = std::find(inputs.begin(), inputs.end(), n->output());
-    if(it != inputs.end()) {
-      size_t p = it - inputs.begin();
-      group->removeInput(p);
-      subgraph.inputs()[p]->replaceAllUsesWith(in_graph->output());
-      subgraph.eraseInput(p);
+    for (size_t i = 0; i < n->outputs().size(); ++i) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/graph_fuser.cpp]: why did this change? multiple outputs can happen due to chunk operators?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 6]
```diff
@@ -436,6 +448,70 @@ struct GraphFuser {
...
+    if (consumer->kind() != prim::FusionGroup) {
+      return false;
+    }
+    // is the output from a chunk node?
+    auto * chunk = producer->node();
+    if (!isChunk(chunk)) {
+      return false;
+    }
+    // Does the chunk have constant chunks/dim?
+    if (!isConstant(chunk->namedInput(attr::chunks)) ||
```
Comment (zdevito) [inline torch/csrc/jit/passes/graph_fuser.cpp]: `chunk->is_constant(attr::chunks)`

**Review by zdevito — COMMENTED** [review]
The fusion pass changes look good. The fusion compiler stuff looks pretty good too but I have some suggestions to simplify it and to avoid allocations along the fast path that I think we should do. Otherwise it would make it difficult to add more functionality to the fuser later.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 5 above)
Comment (zou3519) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Yes, this changed because chunk operators have multiple outputs

File: torch/csrc/jit/fusion_compiler.h (hunk 1 above)
Comment (zou3519) [reply torch/csrc/jit/fusion_compiler.h]: Good point! I'll look into using ConcatDesc and sharing code between chunk/cat

File: torch/csrc/jit/fusion_compiler.cpp [hunk 7]
```diff
@@ -624,41 +668,94 @@ void compressContiguous(
...
+  JIT_ASSERT(chunkDesc.nSubtensors > 1);
+  JIT_ASSERT(sizes[chunkDesc.dim] % chunkDesc.nSubtensors == 0);
+  sizes[chunkDesc.dim] /= chunkDesc.nSubtensors;
+  return sizes;
+}
+
+// XXX: this code assumes that inputs are 32-bit addressable
+static inline uint32_t computeNumel(at::ArrayRef<int64_t> sizes) {
+  uint32_t result = 1;
+  if (sizes.size() == 0) {
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.cpp]: Isn't that the scalar case? Shouldn't it return `1` as numel?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 8]
```diff
@@ -624,41 +668,94 @@ void compressContiguous(
...
   // XXX: this code assumes that all inputs are of the same size
   JIT_ASSERT(inputs[0].numel() <= std::numeric_limits<uint32_t>::max());
-  uint32_t numel = inputs[0].numel();
-  at::IntList map_size = inputs[0].sizes();
+
+  // Compute map_size, numel from the first input
+  at::IntList map_size;
+  uint32_t numel;
+  std::vector<int64_t> keep_alive_size;
+  if (chunk_desc[0].nSubtensors == 1) {
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.cpp]: Didn't you assert this never happens above?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 9]
```diff
@@ -610,6 +682,10 @@ struct GraphFuser {
         any_changed |= changed;
       }
     }
+    // Fuse starting chunks into the group.
+    for (auto it = block->nodes().rbegin(); it != block->nodes().rend();) {
+      it = scanNodeForChunks(*it);
+    }
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Yes I set an invariant that fused chunks can only take in inputs to fusion groups.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 10]
```diff
@@ -445,11 +458,39 @@ std::pair<std::vector<ConcatDesc>, bool> emitCompilationUnit(
...
+  std::vector<std::pair<Value*,TensorDesc&>> flat_inputs;
   {
-    size_t i = 0;
-    for(auto p : subgraph.inputs())
-      emitFormal(p,agraph.input_desc[i++]);
+    size_t input_index = 0;
+    for(auto p : subgraph.inputs()) {
+      if (auto chunk = maybeIsChunk(p)) {
+        int64_t dim = toIValue(chunk->namedInput(attr::dim))->toInt();
+        int64_t chunks = toIValue(chunk->namedInput(attr::chunks))->toInt();
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/jit/fusion_compiler.cpp]: I haven't thought much about it but I will think about it

[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/jit/fusion_compiler.cpp]: Oh I see, `prim::FusedConcat` has `dim` as an attribute on the node. Yes, that does look nicer, I'll give it a shot

**Comment by zou3519** [comment]
This should be ready for another review now, despite the failing (unrelated) tests

File: test/expect/TestJit.test_fusion_distribute.expect [hunk 11]
```diff
@@ -1,20 +1,16 @@
 graph(%0 : Float(4, 4)
       %1 : Float(4, 4)) {
-  %2 : int = prim::Constant[value=1]()
-  %3 : int = prim::Constant[value=2]()
-  %4 : Float(4!, 2), %5 : Float(4!, 2) = aten::chunk(%0, %3, %2)
-  %6 : Float(4!, 2), %7 : Float(4!, 2) = aten::chunk(%1, %3, %2)
-  %8 : Float(4, 2) = prim::FusionGroup_0[device=0](%4, %6, %5, %7)
-  return (%8);
+  %2 : Float(4, 2) = prim::FusionGroup_0[device=0](%0, %1)
```
Comment (apaszke) [inline test/expect/TestJit.test_fusion_distribute.expect]: This test is for distributing chunk over pointwise ops we could fuse. Is that transformation still a thing? We should be able to fuse all chunks anyway, right? Did you remove it?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 12]
```diff
@@ -418,7 +418,18 @@ std::string encodeRHS(Node * n) {
   return format(str, env);
 }
 
-std::pair<std::vector<ConcatDesc>, bool> emitCompilationUnit(
+static at::optional<Node&> usedInFusedChunk(Value * input) {
+  // If input is an input to prim::FusedChunk, it will only have one use.
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.cpp]: Assert plz

File: torch/csrc/jit/fusion_compiler.h [hunk 13]
```diff
@@ -60,24 +60,37 @@ struct AnnotatedGraph {
...
     JIT_ASSERT(nSubtensors > 1);
     std::vector<bool> cont = desc.contiguity;
-    if(dim > 0) {
-      // when we narrow the concatenated output
-      // we make the size[dim] smaller while keeping the stride[dim] the same,
-      // meaning: stride[dim - 1] != stride[dim]*size[dim]
-      // so dim - 1 is no longer contiguous
-      cont[dim - 1] = false;
+    if (is_chunk) {
+      subtensorDesc.reset(new TensorDesc(desc.scalar_type, desc.contiguity));
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.h]: Are you sure that we don't need to modify the contiguity similarly to what we do below? Chunking along any dimension other than the first one will change a contiguous input to non-contiguous chunks.

```python
In [2]: import torch

In [3]: x = torch.randn(10, 10)

In [4]: x.chunk(10, 0)[0].is_contiguous()
Out[4]: True

In [5]: x.chunk(10, 1)[0].is_contiguous()
Out[5]: False
```

**Review by apaszke — APPROVED** [review]
LGTM, but I'm not sure if you're handling contiguity information correctly in `PartitionDesc`

File: test/expect/TestJit.test_fusion_distribute.expect (hunk 11 above)
Comment (zou3519) [reply test/expect/TestJit.test_fusion_distribute.expect]: We still distribute the chunks before we fuse any chunks in. Fusing the chunks in first prevents possible fusion opportunities (because the invariant for prim::FusedChunk is that only inputs to the FusionGroup can be inputs to prim::FusedChunk).

This expect files shows that the chunk was distributed over the fusible pointwise ops, then everything was fused into one fusion group

File: torch/csrc/jit/fusion_compiler.h (hunk 13 above)
Comment (zou3519) [reply torch/csrc/jit/fusion_compiler.h]: Yes, `desc.contiguity` should already have the correct contiguity:
https://github.com/pytorch/pytorch/blob/60b58f153f0325eeb585d030da9af61f124b4aa7/torch/csrc/jit/fusion_compiler.cpp#L470-L476 
That part of the code constructs `desc` with the correct sizes / strides of each of the output chunks, and then the TensorDesc constructor computes the correct contiguity. `desc` is then passed into the PartitionDesc ctor. 

This behavior is different from what happens in a `cat` operation: for `aten::cat` PartitionDesc computes the correct contiguity, but for `aten::chunk` PartitionDesc is passed the correct contiguity. I'll add comments, a separate constructor, or move some code around as appropriate to make this distinction clearer.
